test/failures_testsuite/testcases/s3_redundant/test04_minio_active_passive.py

```diff
@@ -1,23 +1,82 @@
 from base_test import BaseTest
-import requests
 from requests.exceptions import ConnectionError
-import time
-import unittest
+import time, unittest, requests
+from jumpscale import j
+from subprocess import Popen, PIPE
 
 
 class TestActivePassive(BaseTest):
+    @classmethod
+    def setUpClass(cls):
+        """
+        Deploy s3 redundant.
+
+        function to deploy s3 redundant with one of pre-configured parameters.
+
+        """
+        cls.config = j.data.serializer.yaml.load('./config.yaml')
+        if cls.config['s3_redundant']['deploy']:
+            cls.s3_redundant_controller = Controller(cls.config)
+
+            data = [cls.config['s3_redundant']['instance']['farm'], cls.config['s3_redundant']['instance']['size'],
+                    cls.config['s3_redundant']['instance']['shards'], cls.config['s3_redundant']['instance']['parity']]
+
+            for _ in range(5):
+                cls.s3_redundant_service_name = 's3_redundant_{}'.format(str(time.time()).split('.')[0])
+                cls.logger.info("s3 redundant service name : {}".format(cls.s3_redundant_service_name))
+
+                instance = cls.s3_redundant_controller.deploy_s3_redundant(cls.s3_redundant_service_name, *data)
+                try:
+                    cls.logger.info("wait for deploying {} service".format(cls.s3_redundant_service_name))
+                    instance.wait(die=True)
+                    break
+                except Exception as e:
+                    cls.logger.error("There is an error while installing s3 .. we will re-install it!")
+                    cls.logger.error(e)
+                    cls.logger.info('uninstall {} service'.format(cls.s3_redundant_service_name))
+                    s3_redundant_object = cls.s3_redundant_controller.s3_redundant[cls.s3_redundant_service_name]
+                    s3_redundant_object.uninstall()
+                    cls.logger.info('delete {} service'.format(cls.s3_redundant_service_name))
+                    s3_redundant_object.delete()
+            else:
+                raise TimeoutError("can't install s3 redundant .. gonna quit!")
+
+            cls.logger.info('wait for {} state to be okay'.format(cls.s3_redundant_service_name))
+            for _ in range(10):
+                cls.s3_redundant_object = cls.s3_redundant_controller.s3_redundant[cls.s3_redundant_service_name]
+                state = cls.s3_redundant_object.service.state
+                cls.logger.info("{} state : {}".format(cls.s3_redundant_service_name, state))
+                try:
+                    cls.logger.info("waiting {} state to be ok ... ".format(cls.s3_redundant_service_name))
+                    state.check('actions', 'install', 'ok')
+                    break
+                except:
+                    time.sleep(5 * 60)
+                    cls.logger.info("wait for 5 mins .. then we try again!")
+            else:
+                state.check('actions', 'install', 'ok')
```

Comment:
shouldn't you also check if  url['public'] and  in url['storage'] exist for passive and active minios as u did for s3 service in base_test.py?

Reply:
done